src/hotspot/share/gc/shenandoah/shenandoahTaskqueue.hpp

```diff
@@ -181,12 +175,42 @@ class ObjArrayChunkedTask
     return ((uintptr_t) pow) << pow_shift;
   }
 
-  inline oop obj()   const { return decode_oop(_obj);   }
-  inline int chunk() const { return decode_chunk(_obj); }
-  inline int pow()   const { return decode_pow(_obj);   }
-  inline bool is_not_chunked() const { return (_obj & ~right_n_bits(oop_bits + pow_bits)) == 0; }
+public:
+  ShenandoahMarkTask(oop o = NULL) {
+    uintptr_t enc = encode_oop(o);
+    assert(decode_oop(enc) == o,
+           "oop encoding should work: " PTR_FORMAT, p2i(o));
+    assert(decode_not_chunked(enc),
+           "task should not be chunked");
+    _obj = enc;
+  }
+
+  ShenandoahMarkTask(oop o, int chunk, int pow) {
+    uintptr_t enc_oop = encode_oop(o);
+    uintptr_t enc_chunk = encode_chunk(chunk);
+    uintptr_t enc_pow = encode_pow(pow);
+    uintptr_t enc = enc_oop | enc_chunk | enc_pow;
+    assert(decode_oop(enc) == o,
+           "oop encoding should work: " PTR_FORMAT, p2i(o));
+    assert(decode_chunk(enc) == chunk,
+           "chunk encoding should work: %d", chunk);
+    assert(decode_pow(enc) == pow,
+           "pow encoding should work: %d", pow);
+    assert(!decode_not_chunked(enc),
+           "task should be chunked");
+    _obj = enc;
+  }
+
+  // Trivially copyable.
+
+public:
+  inline oop  obj()            const { return decode_oop(_obj);   }
+  inline int  chunk()          const { return decode_chunk(_obj); }
+  inline int  pow()            const { return decode_pow(_obj);   }
+
+  inline bool is_not_chunked() const { return decode_not_chunked(_obj); }
 
-  DEBUG_ONLY(bool is_valid() const); // Tasks to be pushed/popped must be valid.
+  DEBUG_ONLY(bool is_valid() const;) // Tasks to be pushed/popped must be valid.
```

Comment:
That looks weird. I guess I prefer the previous notation, but it's your call.

Reply:
Actually, the previous version looks weird (and CLion highlights it): `DEBUG_ONLY` is a macro that removes all contents, so `;` is left dangling in release mode.